### PR TITLE
appl-riscv.mk build defs

### PR DESCRIPTION
**Pull request by yodada**
In this PR:
 - fix the missing compile time definitions when building within replicant (e.g., parameterized apps)

### CODE REVIEW

**Comment by yodada** [comment]
This PR should receive a second pair of eyes ;)

File: examples/cuda/riscv.mk [hunk 1]
```diff
@@ -73,6 +79,7 @@ NC=\033[0m
 _REPO_ROOT ?= $(shell git rev-parse --show-toplevel)
 -include $(_REPO_ROOT)/environment.mk
 -include $(BSG_MACHINE_PATH)/Makefile.machine.include
+-include $(BSG_MANYCORE_DIR)/software/mk/Makefile.dimensions
```
Comment (mrutt92) [inline examples/cuda/riscv.mk:82]: so we are controlling the tile group dimensions using a makefile in manycore?

Comment (mrutt92) [inline examples/cuda/riscv.mk:82]: so we are controlling the tile group dimensions using a makefile in manycore?

File: examples/cuda/riscv.mk [hunk 2]
```diff
@@ -226,39 +233,55 @@ RISCV_INCLUDES += -I$(BSG_MANYCORE_DIR)/software/bsg_manycore_lib
...
+# endif
+# ifndef BSG_MACHINE_HOST_Y_CORD
+# $(error $(shell echo -e "$(RED)BSG MAKE ERROR: BSG_MACHINE_HOST_Y_CORD undefined"))
+# endif
+# RISCV_DEFINES += -DIO_X_INDEX=$(BSG_MACHINE_HOST_X_CORD)
+# RISCV_DEFINES += -DIO_Y_INDEX=$(BSG_MACHINE_HOST_Y_CORD)
+# RISCV_DEFINES += -DPREALLOCATE=0
+# RISCV_DEFINES += -DHOST_DEBUG=0
+
+RISCV_DEFINES +=-Dbsg_tiles_X=$(bsg_tiles_X) -Dbsg_tiles_Y=$(bsg_tiles_Y)
```
Comment (mrutt92) [inline examples/cuda/riscv.mk:269]: i think these can be derived from the Makefile.machine.include file.

**Comment by mrutt92** [comment]
which defines were missing?

File: examples/cuda/riscv.mk (hunk 2 above)
Comment (yodada) [reply examples/cuda/riscv.mk:269]: I don't really know ;)
I just copied from makefile.builddef

**Comment by yodada** [comment]
if I build with hw barrier it complains about `BARRIER_RUCHE_FACTOR_X` not defined

File: examples/cuda/riscv.mk (hunk 1 above)
Comment (yodada) [reply examples/cuda/riscv.mk:82]: hmmmm. I think this is what happens when building non-parameterized apps? 
https://github.com/cornell-brg/bsg_manycore/blob/63255d758e3bf5c3a55ff5a891f983ac98d11d8a/software/mk/Makefile.builddefs#L17

Comment (yodada) [reply examples/cuda/riscv.mk:82]: So if I get it right ...
https://github.com/cornell-brg/bsg_manycore/blob/63255d758e3bf5c3a55ff5a891f983ac98d11d8a/software/spmd/bsg_cuda_lite_runtime/appl-fib/Makefile#L27
which links
https://github.com/cornell-brg/bsg_manycore/blob/63255d758e3bf5c3a55ff5a891f983ac98d11d8a/software/spmd/Makefile.include#L11
which links
https://github.com/cornell-brg/bsg_manycore/blob/63255d758e3bf5c3a55ff5a891f983ac98d11d8a/software/mk/Makefile.master#L9
which links
https://github.com/cornell-brg/bsg_manycore/blob/63255d758e3bf5c3a55ff5a891f983ac98d11d8a/software/mk/Makefile.builddefs#L17

**Comment by yodada** [comment]
Arguably I'm not that familiar with these Makefiles. The observable problem is when I want to use the hw barrier, it complains about missing `BARRIER_RUCHE_FACTOR_X`

**Comment by mrutt92** [comment]
where are you trying to use the hw barrier?

**Comment by mrutt92** [comment]
can you point me to the code?

**Comment by yodada** [comment]
Oh it's still in a branch.

https://github.com/cornell-brg/bsg_manycore/blob/2014291662c600209af96ea724902d8e4f548853/software/spmd/appl/appl-runtime-applstatic.inl#L13

Wanted to include an `appl::sync` so it can be used in the user code and `applstatic`
https://github.com/cornell-brg/bsg_manycore/blob/2014291662c600209af96ea724902d8e4f548853/software/spmd/appl/appl-hw-barrier.hpp#L33
https://github.com/cornell-brg/bsg_manycore/blob/2014291662c600209af96ea724902d8e4f548853/software/spmd/applstatic/applstatic-parallel_for.inl#L91

**Comment by mrutt92** [comment]
So... in replicant you shouldn't need ruche factor to be defined because the host does barrier setup and the tiles just a read a value from an array and configure the csr
with that value.

**Comment by yodada** [comment]
o....k.... then how do you set the hw barrier up?
I thought you need this?
https://github.com/cornell-brg/bsg_manycore/blob/2014291662c600209af96ea724902d8e4f548853/software/spmd/appl/appl-hw-barrier.cpp#L22-L34

**Comment by mrutt92** [comment]
That's the api you should use if you're running without cuda lite.

Look at examples/cuda/hw_barrier for an example of using it in cuda.

**Comment by yodada** [comment]
looks like some of our changes break `examples/cuda/hw_barrier`
```
Error: "/work/global/lc873/work/sdh/playground/bsg_bladerunner-v6.4-new-profiler/bsg_manycore/v/vanilla_bean/lsu.v", 208: replicant_tb_top.testbench.DUT.py[0].podrow.px[0].pod.mc_y[0].mc_x[0].mc.y[6].x[15].tile.proc.h.z.vcore.lsu0.unnamed$$_1: at time 116519240 ps
[BSG_ERROR] LR operation can only be made on local DMEM address space.
```

**Comment by yodada** [comment]
Tried to use `bsg_barrier_hw_tile_group_sync` and it hangs somehow ... I remember the first time I tried this and it did not work for me. that's why I switched to use the raw API

**Comment by mrutt92** [comment]
on this branch? or on a more up-to-date code base? i've never seen that issue myself.

**Comment by mrutt92** [comment]
the LR-AQ issue comes from using the amoadd barrier to configure. you need a sw barrier of some sort to use before configuring the hw barrier. you would need that even if you were using the raw api.

**Comment by yodada** [comment]
both manycore and replicant on `hb-finegrained-runtime` branch

`examples/cuda/hw_barrie` throws these LR errors

**Comment by mrutt92** [comment]
A simple fix is to edit this function: https://github.com/cornell-brg/bsg_manycore/blob/d30a040ac2d7a136b4216b6e25b0409431d84d69/software/bsg_manycore_lib/bsg_cuda_lite_barrier.h#L15

Allocate sense statically instead of on the stack.

**Comment by yodada** [comment]
and this seems to be working fine?
https://github.com/cornell-brg/bsg_manycore/blob/lc873-05-19-applstatic-merge/software/spmd/appl/appl-hw-barrier.cpp

are you saying some tiles might get to the first sync before all tiles can finish the setup?

**Comment by mrutt92** [comment]
yes, and that works fine because amoadd::alarm is statically allocated and its address is a local pointer.
I'm suggesting you make a one-line edit to the function i pointed you to do the same.

it's written to stack allocate sense because of the assumption that stack pointers are local. your changes break that assumption but it's a very easy fix.

**Comment by yodada** [comment]
Made these changes and `hw_barrier` example works
```
diff --git a/software/bsg_manycore_lib/bsg_cuda_lite_barrier.c b/software/bsg_manycore_lib/bsg_cuda_lite_barrier.c
index 61c81cdf..751febfd 100644
--- a/software/bsg_manycore_lib/bsg_cuda_lite_barrier.c
+++ b/software/bsg_manycore_lib/bsg_cuda_lite_barrier.c
@@ -1 +1,2 @@
  int *__cuda_barrier_cfg;
+ int sense;
diff --git a/software/bsg_manycore_lib/bsg_cuda_lite_barrier.h b/software/bsg_manycore_lib/bsg_cuda_lite_barrier.h
index 58e326f4..7a22194a 100644
--- a/software/bsg_manycore_lib/bsg_cuda_lite_barrier.h
+++ b/software/bsg_manycore_lib/bsg_cuda_lite_barrier.h
@@ -7,6 +7,7 @@
 extern "C" {
 #endif
 extern int *__cuda_barrier_cfg;
+extern int sense;

 /**
  * Initialize the tile-group barrier.
@@ -14,7 +15,7 @@ extern int *__cuda_barrier_cfg;
  */
 static inline void bsg_barrier_hw_tile_group_init()
 {
-    int sense = 1;
+    sense = 1;
     // initalize csr
     int cfg = __cuda_barrier_cfg[1+__bsg_id];
     asm volatile ("csrrw x0, 0xfc1, %0" : : "r" (cfg));
```

**Comment by mrutt92** [comment]
nice. i don't think this pr is required with that change.

**Pull request closed without merging**